### PR TITLE
add metrics stat for microbatch service

### PR DESCRIPTION
**Pull request by hrmthw**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/marshal/marshal.py [hunk 1]
```diff
@@ -203,15 +212,45 @@ def fork_start_app(self, port):
...
+
+    def async_start(self, port):
+        """
+        Start an micro batch server at the specific port on the instance or parameter.
+        """
+        track_server('marshal')
+        marshal_proc = multiprocessing.Process(
+            target=self.fork_start_app, kwargs=dict(port=port), daemon=True,
+        )
+        # TODO: make sure child process dies when parent process is killed.
```
Comment (parano) [inline bentoml/marshal/marshal.py]: use an atexit callback? https://docs.python.org/3/library/atexit.html

File: bentoml/server/marshal_server.py [hunk 2]
```diff
@@ -115,16 +70,34 @@ def load_config(self):
         for key, value in gunicorn_config.items():
             self.cfg.set(key.lower(), value)
 
+    def setup_prometheus_multiproc_dir(self):
```
Comment (parano) [inline bentoml/server/marshal_server.py]: move this function out and share the code with the one in the gunicorn API server?

File: bentoml/server/gunicorn_config.py [hunk 3]
```diff
@@ -54,3 +55,4 @@ def worker_abort(worker):
 def post_worker_init(worker):
     worker.log.debug('Unregistering usage tracking in worker process')
     atexit.unregister(track_server_stop)
+    atexit.unregister(_exit_function)  # Shutting down Gunicorn gracefully
```
Comment (parano) [inline bentoml/server/gunicorn_config.py]: any reason for this? I thought  _exit_function is already in the atexit callback list by default?

File: bentoml/marshal/marshal.py [hunk 4]
```diff
@@ -12,136 +12,159 @@
...
         self.target_host = target_host
         self.target_port = target_port
         self.batch_handlers = dict()
 
+        self.bento_service_metadata_pb = load_bento_service_metadata(bento_bundle_path)
+
+        self.setup_routes_from_pb(self.bento_service_metadata_pb)
+
+        self.CONNECTION_LIMIT = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
+        logger.info(
```
Comment (parano) [inline bentoml/marshal/marshal.py]: 👍

File: bentoml/marshal/marshal.py [hunk 5]
```diff
@@ -203,15 +212,45 @@ def fork_start_app(self, port):
...
+            reqs_s = DataLoader.merge_requests(requests)
+            try:
+                async with aiohttp.ClientSession() as client:
+                    async with client.post(
+                        api_url, data=reqs_s, headers=headers
+                    ) as resp:
+                        raw = await resp.read()
+                merged = DataLoader.split_responses(raw)
+            except aiohttp.ClientConnectorError:
+                return (aiohttp.web.HTTPServiceUnavailable,) * len(requests)
```
Comment (parano) [inline bentoml/marshal/marshal.py]: 👍

File: bentoml/marshal/marshal.py [hunk 6]
```diff
@@ -12,136 +12,159 @@
...
+            self.metrics_request_batch_size.labels(endpoint=api_name).observe(
+                len(requests)
+            )
+            return await func(requests, api_name)
 
-        return _func
+    return _MarshalService
 
 
+@metrics_patch
```
Comment (parano) [inline bentoml/marshal/marshal.py]: I remember we discussed this, do we still want to add a config to disable/enable this?  something like`@metrics_patch(conf['marshal']['enable_metrics'])` here?

**Comment by parano** [comment]
@hrmthw LGTM! merging now, feel free to address comments in follow-up

File: bentoml/server/gunicorn_config.py (hunk 3 above)
Comment (hrmthw) [reply bentoml/server/gunicorn_config.py]: https://github.com/benoitc/gunicorn/issues/1391

**Pull request merged**